### PR TITLE
Label model symmetry breaking

### LINKED ISSUE #1437: negative predictions for data point with positive labels
Hi,
I am using snorkel for a search result quality prediction project.
I have quite a few cases where the predictions do not quite make sense to me.
There are false negative results where the predictions are 0 (negative), but if you look at the labels assigned by each labeling functions, they are either 1(positive) or -1(abstain). There are nearly no 0(negative) at all.
Could anyone help clarify this for me please?
Thanks,
Peng

### PR DESCRIPTION
**Pull request by ajratner**
## Description of proposed changes
This PR primarily implements a heuristic procedure for selecting one of several symmetric (equally optimal) solutions to the `LabelModel` parameter estimation procedure arising from orthogonal symmetries there.  Basically, for any `mu` that we learn (the estimated conditional probabilities of the LFs), we can often also accept column permutations of this result.  So, we choose the solution where the most LFs are estimated to be better than random, as per our standard modeling assumption.

This PR also:
- Slightly refactors and changes the `LabelModel.get_conditional_probs` sub-function
- Factors out the 'post-processing' operations in `LabelModel.fit`, i.e. right now, this symmetry breaking operation and clamping.

## Related issue(s)

Fixes #1437 (at least to first order)

## Test plan
Adding additional tests for (a) conditional probability table calculation, and (b) symmetry breaking specifically

## Checklist
* [x] I have read the **CONTRIBUTING** document.
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.

### CODE REVIEW

File: snorkel/labeling/model/label_model.py [hunk 1]
```diff
@@ -283,52 +283,57 @@ def _init_params(self) -> None:
...
         ----------
-        source
-            Index of source to generate conditional probabilities for, by default None
+        mu
+            An [m * k, k] np.ndarray with entries in [0, 1]
 
         Returns
         -------
         np.ndarray
             Conditional probabilities table if source is None, else corresponding block
```
Comment (paroma) [inline snorkel/labeling/model/label_model.py]: this should change given there's no source input anymore

File: snorkel/labeling/model/label_model.py [hunk 2]
```diff
@@ -283,52 +283,57 @@ def _init_params(self) -> None:
...
+        cprobs = np.zeros((self.m, self.cardinality + 1, self.cardinality))
         for i in range(self.m):
             # si = self.c_data[(i,)]['start_index']
             # ei = self.c_data[(i,)]['end_index']
             # mu_i = mu[si:ei, :]
             mu_i = mu[i * self.cardinality : (i + 1) * self.cardinality, :]
-            c_probs[
-                i * (self.cardinality + 1) + 1 : (i + 1) * (self.cardinality + 1), :
-            ] = mu_i
+            cprobs[i, 1:, :] = mu_i
```
Comment (paroma) [inline snorkel/labeling/model/label_model.py]: nit: add comment this is skipping the abstain?

File: snorkel/labeling/model/label_model.py [hunk 3]
```diff
@@ -283,52 +283,57 @@ def _init_params(self) -> None:
...
+
+            cprobs[i, j, k] = P(\lf_i = j-1 | Y = k)
+
+        where m is the number of LFs, k is the cardinality, and cprobs includes the
+        conditional abstain probabilities P(\lf_i = -1 | Y = y).
+
+        Returns
+        -------
+        np.ndarray
+            Conditional probabilities table if source is None, else corresponding block
```
Comment (paroma) [inline snorkel/labeling/model/label_model.py]: source is not an input

File: snorkel/labeling/model/label_model.py [hunk 4]
```diff
@@ -283,52 +283,57 @@ def _init_params(self) -> None:
...
+            cprobs[i, 0, :] = 1 - mu_i.sum(axis=0)
+        return cprobs
 
-        if source is not None:
-            return c_probs[
-                source * (self.cardinality + 1) : (source + 1) * (self.cardinality + 1)
-            ]
-        else:
-            return c_probs
+    def get_conditional_probs(self) -> np.ndarray:
```
Comment (paroma) [inline snorkel/labeling/model/label_model.py]: do we want this to be a public method to let users debug?

File: snorkel/labeling/model/label_model.py [hunk 5]
```diff
@@ -693,6 +697,71 @@ def _update_lr_scheduler(self, step: int) -> None:
...
+        If there are several values of mu that equivalently satisfy the optimization
+        objective, as there often are due to column permutation symmetries, then pick
+        the solution that trusts the user-written LFs most.
+
+        In more detail, suppose that mu satisfies (minimizes) the two loss objectives:
+            1. O = mu @ P @ mu.T
+            2. diag(O) = sum(mu @ P, axis=1)
+        Then any column permutation matrix Z that commutes with P will also equivalently
+        satisfy these objectives, and thus is an equally valid (symmetric) solution.
+        Therefore, we select the solution where the most LFs are estimated to have more
```
Comment (paroma) [inline snorkel/labeling/model/label_model.py]: super nit: more --> higher

File: snorkel/labeling/model/label_model.py (hunk 1 above)
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: ?  I deleted that?

File: snorkel/labeling/model/label_model.py (hunk 4 above)
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: Yeah!

[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: Yes- this is a pretty basic output from the `LabelModel`

File: snorkel/labeling/model/label_model.py [hunk 6]
```diff
@@ -347,10 +352,9 @@ def get_weights(self) -> np.ndarray:
         array([0.99, 0.99, 0.99])
         """
         accs = np.zeros(self.m)
```
[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: huh?  These are still the accuracies here

File: snorkel/labeling/model/label_model.py [hunk 7]
```diff
@@ -693,6 +697,71 @@ def _update_lr_scheduler(self, step: int) -> None:
...
+        Then any column permutation matrix Z that commutes with P will also equivalently
+        satisfy these objectives, and thus is an equally valid (symmetric) solution.
+        Therefore, we select the solution where the most LFs are estimated to have more
+        net probability of being correct, P(\lf = y, Y = y), than incorrect, which is
+        the standard assumption we have made to date.
+
+        Note that this is not the only possible heuristic / assumption that we could
+        use, and in practice this may require further iteration.
+        """
+        mu = self.mu.detach().clone().numpy()
```
[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: 🤷‍♂

[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: Well I think `detach().clone()` --> `detach()`.  Need to see if this is still needed with `numpy()` though

[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: Actually these are all read-only so going to just do `.numpy()`

[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/labeling/model/label_model.py]: And nvm... `RuntimeError: Can't call numpy() on Variable that requires grad. Use var.detach().numpy() instead`.

So `.detach().numpy()` it is!

File: snorkel/labeling/model/label_model.py [hunk 8]
```diff
@@ -693,6 +697,96 @@ def _update_lr_scheduler(self, step: int) -> None:
...
+        this rounding is done to make it more obvious when the parameters have been
+        clamped.
+        """
+        if self.train_config.mu_eps is not None:
+            mu_eps = self.train_config.mu_eps
+        else:
+            mu_eps = min(0.01, 1 / 10 ** np.ceil(np.log10(self.n)))
+        self.mu.data = self.mu.clamp(mu_eps, 1 - mu_eps)  # type: ignore
+
+    def _count_accurate_LFs(self, mu: np.ndarray) -> int:
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: `_count_accurate_lfs`

File: test/labeling/model/test_label_model.py [hunk 9]
```diff
@@ -352,7 +352,35 @@ def test_set_mu_eps(self):
         L = np.array([[1, 1, 1], [1, 1, 1]])
         label_model = LabelModel(verbose=False)
         label_model.fit(L, mu_eps=mu_eps)
-        self.assertAlmostEqual(label_model._get_conditional_probs(0)[1, 0], mu_eps)
+        self.assertAlmostEqual(label_model.get_conditional_probs()[0, 1, 0], mu_eps)
+
+    def test_count_accurate_LFs(self):
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: `test_count_accurate_lfs`

**Comment by plison** [comment]
I'm experiencing problems with the symmetric breaking code.  Given the combinatorial explosion of the number of possible permutations (as a function of the number of output classes), the method does not really scale to problems with more than 6-7 classes.  Maybe the code should stop after a few thousand permutations?

**Pull request merged**